File: backend/src/test/kotlin/flagent/service/EvalResultTest.kt
package flagent.service

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlinx.serialization.json.*

class EvalResultTest {
    @Test
    fun `create EvalResult with all fields`() {
        val entityContext = buildJsonObject {
            put("region", "US")
            put("tier", "premium")
        }
        
        val result = EvalResult(
            flagID = 1,
            flagKey = "test_flag",
            flagSnapshotID = 5,
            flagTags = listOf("production", "feature"),
            segmentID = 10,
            variantID = 20,
            variantKey = "control",
            variantAttachment = null,
            evalContext = EvalContext(
                entityID = "user123",
                entityType = "user",
                entityContext = entityContext
            ),
            evalDebugLog = null,
            timestamp = 1234567890L
        )
        
        assertEquals(1, result.flagID)
        assertEquals("test_flag", result.flagKey)
        assertEquals(5, result.flagSnapshotID)
        assertEquals(2, result.flagTags.size)
        assertEquals(10, result.segmentID)
        assertEquals(20, result.variantID)
        assertEquals("control", result.variantKey)
        assertEquals("user123", result.evalContext.entityID)
        assertEquals("user", result.evalContext.entityType)
        assertEquals(1234567890L, result.timestamp)
    }
    
    @Test
    fun `create EvalResult with default values`() {
        val result = EvalResult(
            flagID = 1,
            flagKey = "test_flag",
            flagSnapshotID = 1,
            flagTags = emptyList(),
            evalContext = EvalContext(),
            timestamp = 0L
        )
        
        assertNull(result.segmentID)
        assertNull(result.variantID)
        assertNull(result.variantKey)
        assertNull(result.variantAttachment)
        assertNull(result.evalDebugLog)
        assertNull(result.evalContext.entityID)
        assertNull(result.evalContext.entityType)
        assertNull(result.evalContext.entityContext)
    }
    
    @Test
    fun `create EvalResult with variant attachment`() {
        val attachment = buildJsonObject {
            put("color", "red")
            put("enabled", true)
        }
        
        val result = EvalResult(
            flagID = 1,
            flagKey = "test_flag",
            flagSnapshotID = 1,
            flagTags = emptyList(),
            variantID = 1,
            variantKey = "treatment",
            variantAttachment = attachment,
            evalContext = EvalContext(),
            timestamp = 0L
        )
        
        assertEquals(attachment, result.variantAttachment)
        assertEquals("red", result.variantAttachment?.get("color")?.jsonPrimitive?.content)
    }
    
    @Test
    fun `EvalResult data class equality`() {
        val result1 = EvalResult(
            flagID = 1,
            flagKey = "test",
            flagSnapshotID = 1,
            flagTags = emptyList(),
            evalContext = EvalContext(entityID = "user1"),
            timestamp = 1000L
        )
        val result2 = EvalResult(
            flagID = 1,
            flagKey = "test",
            flagSnapshotID = 1,
            flagTags = emptyList(),
            evalContext = EvalContext(entityID = "user1"),
            timestamp = 1000L
        )
        
        assertEquals(result1, result2)
        assertEquals(result1.hashCode(), result2.hashCode())
    }
}

File: sdk/dart/test/eval_result_test.dart
import 'package:built_collection/built_collection.dart';
import 'package:built_value/json_object.dart';
import 'package:test/test.dart';
import 'package:flagent_client/flagent_client.dart';

// tests for EvalResult
void main() {
  final evalCtx = EvalContext((b) => b
    ..entityID = 'user-1'
    ..entityType = 'user'
    ..flagKey = 'test_flag');
  final instance = EvalResult((b) => b
    ..flagID = 1
    ..flagKey = 'test_flag'
    ..flagSnapshotID = 10
    ..flagTags = ListBuilder<String>(['tag1'])
    ..segmentID = 5
    ..variantID = 2
    ..variantKey = 'control'
    ..variantAttachment = MapBuilder<String, JsonObject?>()
    ..evalContext = evalCtx.toBuilder()
    ..timestamp = DateTime.utc(2024, 1, 1, 12, 0, 0)
    ..evalDebugLog = null);

  group(EvalResult, () {
    test('to test the property `flagID`', () async {
      expect(instance.flagID, equals(1));
    });

    test('to test the property `flagKey`', () async {
      expect(instance.flagKey, equals('test_flag'));
    });

    test('to test the property `flagSnapshotID`', () async {
      expect(instance.flagSnapshotID, equals(10));
    });

    test('to test the property `flagTags`', () async {
      expect(instance.flagTags, isNotNull);
      expect(instance.flagTags!.length, equals(1));
      expect(instance.flagTags![0], equals('tag1'));
    });

    test('to test the property `segmentID`', () async {
      expect(instance.segmentID, equals(5));
    });

    test('to test the property `variantID`', () async {
      expect(instance.variantID, equals(2));
    });

    test('to test the property `variantKey`', () async {
      expect(instance.variantKey, equals('control'));
    });

    test('to test the property `variantAttachment`', () async {
      expect(instance.variantAttachment, isNotNull);
      expect(instance.variantAttachment!.isEmpty, isTrue);
    });

    test('to test the property `evalContext`', () async {
      expect(instance.evalContext, isNotNull);
      expect(instance.evalContext!.entityID, equals('user-1'));
      expect(instance.evalContext!.flagKey, equals('test_flag'));
    });

    test('to test the property `timestamp`', () async {
      expect(instance.timestamp, equals(DateTime.utc(2024, 1, 1, 12, 0, 0)));
    });

    test('to test the property `evalDebugLog`', () async {
      expect(instance.evalDebugLog, isNull);
    });
  });
}
